<a href="https://colab.research.google.com/github/adas754/da_hackathon_3/blob/main/case_study_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Defining Problem Statement and Analysing basic metrics

# Problem Statement -> Analyze the data and generate insights that could help Netflix in deciding which type of shows/movies to produce and how they can grow the business in different countries.


# Loading the data and Basic Observations

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv('/content/net.csv')
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...

# 2. Observations on the shape of data, data types of all the attributes, conversion of categorical attributes to 'category'

In [6]:
df.shape

(8807, 12)

total 8807 rows and 12 columns.

#check how many categorical and numerical value is there

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


only release_year is the numerical

#check missing values

In [8]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [14]:
df[df['duration'].isna()]

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year  rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  74 min   
5794  Louis C.K.  United States  September 16, 2016          2010  84 min   
5813  Louis C.K.  United States     August 15, 2016          2015  66 min   

     duration listed_in                                        description  
5541      NaN    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794      NaN    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813      NaN    Movies  The comic puts his trademark hilarious/thought...

3 missing value is there , want to fill it

In [15]:
ind = df[df['duration'].isna()].index

In [16]:
ind

Index([5541, 5794, 5813], dtype='int64')

In [17]:
df.loc[ind] = df.loc[ind].fillna(method = 'ffill' , axis = 1)

<ipython-input-17-89b4f0c8704c>:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.loc[ind] = df.loc[ind].fillna(method = 'ffill' , axis = 1)
<ipython-input-17-89b4f0c8704c>:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[2017 2010 2015]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[ind] = df.loc[ind].fillna(method = 'ffill' , axis = 1)


#statistical summary

In [9]:
df.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

In [10]:
df.describe(include = object)

show_id   type                 title       director  \
count     8807   8807                  8807           6173   
unique    8807      2                  8807           4528   
top         s1  Movie  Dick Johnson Is Dead  Rajiv Chilaka   
freq         1   6131                     1             19   

                      cast        country       date_added rating  duration  \
count                 7982           7976             8797   8803      8804   
unique                7692            748             1767     17       220   
top     David Attenborough  United States  January 1, 2020  TV-MA  1 Season   
freq                    19           2818              109   3207      1793   

                           listed_in  \
count                           8807   
unique                           514   
top     Dramas, International Movies   
freq                             362   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4

#how many unique value

In [11]:
df.nunique()

show_id         8807
type               2
title           8807
director        4528
cast            7692
country          748
date_added      1767
release_year      74
rating            17
duration         220
listed_in        514
description     8775
dtype: int64

#find null value in rating column

In [19]:
df[df.rating.isna()]

show_id     type                                              title  \
5989   s5990    Movie  13TH: A Conversation with Oprah Winfrey & Ava ...   
6827   s6828  TV Show                  Gargantia on the Verdurous Planet   
7312   s7313  TV Show                                       Little Lunch   
7537   s7538    Movie                               My Honor Was Loyalty   

             director                                               cast  \
5989              NaN                        Oprah Winfrey, Ava DuVernay   
6827              NaN  Kaito Ishikawa, Hisako Kanemoto, Ai Kayano, Ka...   
7312              NaN  Flynn Curry, Olivia Deeble, Madison Lu, Oisín...   
7537  Alessandro Pepe  Leone Frisa, Paolo Vaccarino, Francesco Miglio...   

        country        date_added release_year rating  duration  \
5989        NaN  January 26, 2017         2017    NaN    37 min   
6827      Japan  December 1, 2016         2013    NaN  1 Season   
7312  Australia  February 1, 2018         2015    NaN  1 Season   
7537      Italy     March 1, 2017         2015    NaN   115 min   

                                 listed_in  \
5989                                Movies   
6827  Anime Series, International TV Shows   
7312                 Kids' TV, TV Comedies   
7537                                Dramas   

                                            description  
5989  Oprah Winfrey sits down with director Ava DuVe...  
6827  After falling through a wormhole, a space-dwel...  
7312  Adopting a child's perspective, this show take...  
7537  Amid the chaos and horror of World War II, a c...

In [20]:
indices = df[df.rating.isna()].index
indices

Index([5989, 6827, 7312, 7537], dtype='int64')

In [21]:
df.loc[indices , 'rating'] = 'Not Available'

In [23]:
df.rating.unique()

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', '74 min', '84 min', '66 min', 'NR',
       'Not Available', 'TV-Y7-FV', 'UR'], dtype=object)

In rating column , NR (Not rated) is same as UR (Unrated). lets change UR to NR.

In [24]:
df.loc[df['rating'] == 'UR' , 'rating'] = 'NR'

In [29]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country          830
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

# droped data_added column

In [28]:
df.drop(df.loc[df['date_added'].isna()].index , axis = 0 , inplace = True)

In [30]:
df['date_added'].value_counts()

date_added
January 1, 2020      109
November 1, 2019      89
March 1, 2018         75
December 31, 2019     74
October 1, 2018       71
                    ... 
December 4, 2016       1
November 21, 2016      1
November 19, 2016      1
November 17, 2016      1
January 11, 2020       1
Name: count, Length: 1767, dtype: int64

In [33]:
df['date_added'] = pd.to_datetime(df['date_added'].str.strip(), format='%B %d, %Y', errors='coerce')
df['date_added']

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8797, dtype: datetime64[ns]

In [34]:
df['year_added'] = df['date_added'].dt.year

In [35]:
df['month_added'] = df['date_added'].dt.month

In [36]:
df[['date_added' , 'year_added' , 'month_added']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 8797 entries, 0 to 8806
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date_added   8797 non-null   datetime64[ns]
 1   year_added   8797 non-null   int32         
 2   month_added  8797 non-null   int32         
dtypes: datetime64[ns](1), int32(2)
memory usage: 206.2 KB


total null values in each column

In [37]:
df.isna().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country          830
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
year_added         0
month_added        0
dtype: int64

In [38]:
round((df.isna().sum()/ df.shape[0])*100)

show_id          0.0
type             0.0
title            0.0
director        30.0
cast             9.0
country          9.0
date_added       0.0
release_year     0.0
rating           0.0
duration         0.0
listed_in        0.0
description      0.0
year_added       0.0
month_added      0.0
dtype: float64

# 3. Non-Graphical Analysis: Value counts and unique attributes

 2 types of content present in dataset

In [39]:
df['type'].unique()


array(['Movie', 'TV Show'], dtype=object)

In [40]:
movies  = df.loc[df['type'] == 'Movie']
tv_shows = df.loc[df['type'] == 'TV Show']

In [41]:
movies.duration.value_counts()


duration
90 min     152
94 min     146
97 min     146
93 min     146
91 min     144
          ... 
208 min      1
5 min        1
16 min       1
186 min      1
191 min      1
Name: count, Length: 205, dtype: int64

In [42]:
tv_shows.duration.value_counts()

duration
1 Season      1793
2 Seasons      421
3 Seasons      198
4 Seasons       94
5 Seasons       64
6 Seasons       33
7 Seasons       23
8 Seasons       17
9 Seasons        9
10 Seasons       6
13 Seasons       2
15 Seasons       2
12 Seasons       2
17 Seasons       1
11 Seasons       1
Name: count, dtype: int64

In [43]:
movies['duration'] = movies['duration'].str[:-3]
movies['duration'] = movies['duration'].astype('float')

<ipython-input-43-53ddd658a0dd>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['duration'] = movies['duration'].str[:-3]
<ipython-input-43-53ddd658a0dd>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['duration'] = movies['duration'].astype('float')


In [44]:
movies['duration']

0        90.0
6        91.0
7       125.0
9       104.0
12      127.0
        ...  
8801     96.0
8802    158.0
8804     88.0
8805     88.0
8806    111.0
Name: duration, Length: 6131, dtype: float64

In [45]:
tv_shows['duration'] = tv_shows.duration.str[:-7].apply(lambda x : x.strip())
tv_shows['duration'] = tv_shows['duration'].astype('float')

<ipython-input-45-8490b01afea6>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tv_shows['duration'] = tv_shows.duration.str[:-7].apply(lambda x : x.strip())
<ipython-input-45-8490b01afea6>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tv_shows['duration'] = tv_shows['duration'].astype('float')


In [46]:
tv_shows.rename({'duration': 'duration_in_seasons'} ,axis = 1 , inplace = True)
movies.rename({'duration': 'duration_in_minutes'} ,axis = 1 , inplace = True)

<ipython-input-46-6fba49e9528a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tv_shows.rename({'duration': 'duration_in_seasons'} ,axis = 1 , inplace = True)
<ipython-input-46-6fba49e9528a>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies.rename({'duration': 'duration_in_minutes'} ,axis = 1 , inplace = True)


In [47]:
tv_shows.duration_in_seasons

1       2.0
2       1.0
3       1.0
4       2.0
5       1.0
       ... 
8795    2.0
8796    2.0
8797    3.0
8800    1.0
8803    2.0
Name: duration_in_seasons, Length: 2666, dtype: float64

In [48]:
movies.duration_in_minutes

0        90.0
6        91.0
7       125.0
9       104.0
12      127.0
        ...  
8801     96.0
8802    158.0
8804     88.0
8805     88.0
8806    111.0
Name: duration_in_minutes, Length: 6131, dtype: float64

when was first movie added on netflix and when is the most recent movie added on netflix

In [49]:
timeperiod = pd.Series((df['date_added'].min().strftime('%B %Y') , df['date_added'].max().strftime('%B %Y')))
timeperiod.index = ['first' , 'Most Recent']
timeperiod

first            January 2008
Most Recent    September 2021
dtype: object

The oldest and the most recent movie/TV show released on the Netflix in which year

In [50]:
df.release_year.min() , df.release_year.max()


(1925, 2021)

In [51]:
df.loc[(df.release_year == df.release_year.min()) | (df.release_year == df.release_year.max())].sort_values('release_year')

show_id     type                                         title  \
4250   s4251  TV Show             Pioneers: First Women Filmmakers*   
966     s967    Movie                                 Get the Grift   
967     s968  TV Show                      Headspace Guide to Sleep   
968     s969  TV Show                                        Sexify   
972     s973  TV Show                                         Fatma   
...      ...      ...                                           ...   
466     s467  TV Show                            My Unorthodox Life   
467     s468    Movie  Private Network: Who Killed Manuel Buendía?   
468     s469    Movie               The Guide to the Perfect Family   
471     s472    Movie                                Day of Destiny   
8437   s8438  TV Show                        The Netflix Afterparty   

                    director  \
4250                     NaN   
966            Pedro Antonio   
967                      NaN   
968                      NaN   
972                      NaN   
...                      ...   
466                      NaN   
467           Manuel Alcalá   
468            Ricardo Trogi   
471   Akay Mason, Abosi Ogba   
8437                     NaN   

                                                   cast        country  \
4250                                                NaN            NaN   
966   Marcus Majella, Samantha Schmütz, Caito Maini...         Brazil   
967                                 Evelyn Lewis Prieto            NaN   
968   Aleksandra Skraba, Maria Sobocińska, Sandra D...         Poland   
972   Burcu Biricik, Uğur Yücel, Mehmet Yılmaz Ak...         Turkey   
...                                                 ...            ...   
466                                                 NaN            NaN   
467                               Daniel Giménez Cacho            NaN   
468   Louis Morissette, Émilie Bierre, Catherine Ch...            NaN   
471   Olumide Oworu, Denola Grey, Gbemi Akinlade, Ji...            NaN   
8437       David Spade, London Hughes, Fortune Feimster  United States   

     date_added release_year rating  duration  \
4250 2018-12-30         1925  TV-14  1 Season   
966  2021-04-28         2021  TV-MA    95 min   
967  2021-04-28         2021   TV-G  1 Season   
968  2021-04-28         2021  TV-MA  1 Season   
972  2021-04-27         2021  TV-MA  1 Season   
...         ...          ...    ...       ...   
466  2021-07-14         2021  TV-MA  1 Season   
467  2021-07-14         2021  TV-MA   100 min   
468  2021-07-14         2021  TV-MA   102 min   
471  2021-07-13         2021  TV-PG   110 min   
8437 2021-01-02         2021  TV-MA  1 Season   

                                              listed_in  \
4250                                           TV Shows   
966                      Comedies, International Movies   
967                     Docuseries, Science & Nature TV   
968      International TV Shows, TV Comedies, TV Dramas   
972     International TV Shows, TV Dramas, TV Thrillers   
...                                                 ...   
466                                          Reality TV   
467                 Documentaries, International Movies   
468              Comedies, Dramas, International Movies   
471   Children & Family Movies, Dramas, Internationa...   
8437          Stand-Up Comedy & Talk Shows, TV Comedies   

                                            description  year_added  \
4250  This collection restores films from women who ...        2018   
966   After a botched scam, Clóvis bumps into Lohan...        2021   
967   Learn how to sleep better with Headspace. Each...        2021   
968   To build an innovative sex app and win a tech ...        2021   
972   Reeling from tragedy, a nondescript house clea...        2021   
...                                                 ...         ...   
466   Follow Julia Haart, Elite World Group CEO and ...        2021   
467   A deep


Which are different ratings available on Netflix in each type of content



In [52]:
df.groupby(['type' , 'rating'])['show_id'].count()

type     rating       
Movie    66 min              1
         74 min              1
         84 min              1
         G                  41
         NC-17               3
         NR                 78
         Not Available       2
         PG                287
         PG-13             490
         R                 797
         TV-14            1427
         TV-G              126
         TV-MA            2062
         TV-PG             540
         TV-Y              131
         TV-Y7             139
         TV-Y7-FV            5
TV Show  NR                  4
         Not Available       2
         R                   2
         TV-14             730
         TV-G               94
         TV-MA            1143
         TV-PG             321
         TV-Y              175
         TV-Y7             194
         TV-Y7-FV            1
Name: show_id, dtype: int64

In [54]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country          830
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
year_added         0
month_added        0
dtype: int64

In [53]:
df['country'].value_counts()

country
United States                             2812
India                                      972
United Kingdom                             418
Japan                                      244
South Korea                                199
                                          ... 
Romania, Bulgaria, Hungary                   1
Uruguay, Guatemala                           1
France, Senegal, Belgium                     1
Mexico, United States, Spain, Colombia       1
United Arab Emirates, Jordan                 1
Name: count, Length: 748, dtype: int64

In [55]:
country_tb = df[['show_id' , 'type' , 'country']]
country_tb.dropna(inplace = True)
country_tb['country'] = country_tb['country'].apply(lambda x : x.split(','))
country_tb = country_tb.explode('country')
country_tb

<ipython-input-55-88f820136e36>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  country_tb.dropna(inplace = True)
<ipython-input-55-88f820136e36>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  country_tb['country'] = country_tb['country'].apply(lambda x : x.split(','))


show_id     type        country
0         s1    Movie  United States
1         s2  TV Show   South Africa
4         s5  TV Show          India
7         s8    Movie  United States
7         s8    Movie          Ghana
...      ...      ...            ...
8801   s8802    Movie         Jordan
8802   s8803    Movie  United States
8804   s8805    Movie  United States
8805   s8806    Movie  United States
8806   s8807    Movie          India

[10010 rows x 3 columns]

In [56]:
# some duplicate values are found, which have unnecessary spaces. some empty strings found
country_tb['country'] = country_tb['country'].str.strip()

In [57]:
country_tb.loc[country_tb['country'] == '']

show_id     type country
193     s194  TV Show        
365     s366    Movie        
1192   s1193    Movie        
2224   s2225    Movie        
4653   s4654    Movie        
5925   s5926    Movie        
7007   s7008    Movie

In [58]:
country_tb = country_tb.loc[country_tb['country'] != '']

In [59]:
country_tb['country'].nunique()

122